File: mf2/swift/MessageFormat2/Sources/MessageFormat2/RelativeTimeCore.swift
import Foundation

public enum MF2RelativeTimeCore {
    private static let maxLocaleLength = 256
    private static let maxOptionLength = 256
    private static let maxOperandLength = 256
    private static let maxRelativeTimeQuantity = 1_000_000_000.0
    private static let decimalNumberPattern = #"^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$"#

    public enum Style {
        case long
        case short
        case narrow

        public static func fromName(_ value: String) throws -> Style {
            switch try MF2RelativeTimeCore.optionName(value, name: "style") {
            case "long":
                .long
            case "short":
                .short
            case "narrow":
                .narrow
            default:
                throw MF2Error.badOption("style must be one of long, short, narrow.")
            }
        }

        var key: String {
            switch self {
            case .long:
                "long"
            case .short:
                "short"
            case .narrow:
                "narrow"
            }
        }
    }

    public enum Numeric {
        case always
        case auto

        public static func fromName(_ value: String) throws -> Numeric {
            switch try MF2RelativeTimeCore.optionName(value, name: "numeric") {
            case "always":
                .always
            case "auto":
                .auto
            default:
                throw MF2Error.badOption("numeric must be one of always, auto.")
            }
        }
    }

    public enum Policy {
        case precise
        case compact
        case chat

        public static func fromName(_ value: String) throws -> Policy {
            switch try MF2RelativeTimeCore.optionName(value, name: "policy") {
            case "precise":
                .precise
            case "compact":
                .compact
            case "chat":
                .chat
            default:
                throw MF2Error.badOption("policy must be one of precise, compact, chat.")
            }
        }

        var steps: [PolicyStep] {
            switch self {
            case .precise:
                [
                    PolicyStep(upper: 60, unit: .second),
                    PolicyStep(upper: 3_600, unit: .minute),
                    PolicyStep(upper: 86_400, unit: .hour),
                    PolicyStep(upper: 604_800, unit: .day),
                    PolicyStep(upper: 2_592_000, unit: .week),
                    PolicyStep(upper: 31_536_000, unit: .month),
                    PolicyStep(upper: .infinity, unit: .year),
                ]
            case .compact:
                [
                    PolicyStep(upper: 60, unit: .second),
                    PolicyStep(upper: 3_600, unit: .minute),
                    PolicyStep(upper: 86_400, unit: .hour),
                    PolicyStep(upper: .infinity, unit: .day),
                ]
            case .chat:
                [
                    PolicyStep(upper: 45, unit: .second),
                    PolicyStep(upper: 2_700, unit: .minute),
                    PolicyStep(upper: 79_200, unit: .hour),
                    PolicyStep(upper: 604_800, unit: .day),
                    PolicyStep(upper: .infinity, unit: .week),
                ]
            }
        }
    }

    public enum Unit {
        case auto
        case second
        case minute
        case hour
        case day
        case week
        case month
        case quarter
        case year

        public static func fromName(_ value: String) throws -> Unit {
            switch try MF2RelativeTimeCore.optionName(value, name: "unit") {
            case "auto":
                .auto
            case "second":
                .second
            case "minute":
                .minute
            case "hour":
                .hour
            case "day":
                .day
            case "week":
                .week
            case "month":
                .month
            case "quarter":
                .quarter
            case "year":
                .year
            default:
                throw MF2Error.badOption("unit must be one of auto, second, minute, hour, day, week, month, quarter, year.")
            }
        }

        var key: String {
            switch self {
            case .auto:
                "auto"
            case .second:
                "second"
            case .minute:
                "minute"
            case .hour:
                "hour"
            case .day:
                "day"
            case .week:
                "week"
            case .month:
                "month"
            case .quarter:
                "quarter"
            case .year:
                "year"
            }
        }

        var seconds: Double {
            switch self {
            case .auto:
                1
            case .second:
                1
            case .minute:
                60
            case .hour:
                3_600
            case .day:
                86_400
            case .week:
                604_800
            case .month:
                2_592_000
            case .quarter:
                7_776_000
            case .year:
                31_536_000
            }
        }
    }

    public struct Options {
        public var locale: String
        public var style: Style
        public var numeric: Numeric
        public var policy: Policy
        public var unit: Unit

        public init(
            locale: String = "en",
            style: Style = .short,
            numeric: Numeric = .always,
            policy: Policy = .precise,
            unit: Unit = .auto
        ) {
            self.locale = locale
            self.style = style
            self.numeric = numeric
            self.policy = policy
            self.unit = unit
        }
    }

    public struct Data: Decodable {
        public let localeMap: [String: String]
        public let patternSets: [PatternSet]

        public init(localeMap: [String: String], patternSets: [PatternSet]) {
            self.localeMap = localeMap
            self.patternSets = patternSets
        }

        private enum CodingKeys: String, CodingKey {
            case localeMap
            case patternSets
        }
    }

    public struct PatternSet: Decodable {
        public let id: String
        public let data: [String: [String: UnitData]]

        public init(id: String, data: [String: [String: UnitData]]) {
            self.id = id
            self.data = data
        }
    }

    public struct UnitData: Decodable {
        public let future: [String: String]
        public let past: [String: String]
        public let relative: [String: String]

        public init(
            future: [String: String] = [:],
            past: [String: String] = [:],
            relative: [String: String] = [:]
        ) {
            self.future = future
            self.past = past
            self.relative = relative
        }

        private enum CodingKeys: String, CodingKey {
            case future
            case past
            case relative
        }

        public init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            future = try container.decodeIfPresent([String: String].self, forKey: .future) ?? [:]
            past = try container.decodeIfPresent([String: String].self, forKey: .past) ?? [:]
            relative = try container.decodeIfPresent([String: String].self, forKey: .relative) ?? [:]
        }
    }

    public struct Formatter {
        private let localeMap: [String: String]
        private let patternSets: [String: PatternSet]

        public init(data: Data) throws {
            guard !data.localeMap.isEmpty, !data.patternSets.isEmpty else {
                throw MF2Error.missingLocaleData("Relative-time core data has an unsupported shape.")
            }
            localeMap = data.localeMap
            var nextPatternSets: [String: PatternSet] = [:]
            for patternSet in data.patternSets where !patternSet.id.isEmpty && !patternSet.data.isEmpty {
                nextPatternSets[patternSet.id] = patternSet
            }
            patternSets = nextPatternSets
            if patternSets.isEmpty {
                throw MF2Error.missingLocaleData("Relative-time core data has an unsupported shape.")
            }
        }

        public func registry() -> MF2FunctionRegistry {
            MF2FunctionRegistry.portable
                .withFunction("relativeTime") { call in
                    try formatCall(call)
                }
        }

        public func format(_ value: MF2Value, options: Options = Options()) throws -> String {
            let seconds = try parseFiniteNumber(value)
            let unit = options.unit == .auto ? selectUnit(seconds, policy: options.policy) : options.unit
            let quantity = try relativeTimeQuantity(seconds, unit: unit)
            let locale = try MF2RelativeTimeCore.localeOption(options.locale)

            if useRelativeZero(policy: options.policy, numeric: options.numeric, seconds: seconds),
               let relative = try relativeTerm(locale: locale, style: options.style, unit: unit, offset: "0")
            {
                return relative
            }
            if options.numeric == .auto, let offset = relativeOffset(seconds, unit: unit, quantity: quantity),
               let relative = try relativeTerm(locale: locale, style: options.style, unit: unit, offset: offset)
            {
                return relative
            }

            let direction = isNegativeRelativeTime(seconds) ? "past" : "future"
            let data = try unitData(locale: locale, style: options.style, unit: unit)
            let category = NumberOperands(quantity.text)
                .map { selectCardinal(locale: locale, operands: $0) } ?? "other"
            let pattern = try relativeTimePattern(
                data: data,
                locale: locale,
                style: options.style,
                unit: unit,
                direction: direction,
                category: category
            )
            return pattern.replacingOccurrences(of: "{0}", with: quantity.text)
        }

        public func formatToParts(_ value: MF2Value, options: Options = Options()) throws -> [MF2FormattedPart] {
            [.text(try format(value, options: options))]
        }

        private func formatCall(_ call: MF2FunctionCall) throws -> String {
            let options = try Options(
                locale: call.locale,
                style: Style.fromName(try call.optionValue("style", default: "short") ?? "short"),
                numeric: Numeric.fromName(try call.optionValue("numeric", default: "always") ?? "always"),
                policy: Policy.fromName(try call.optionValue("policy", default: "precise") ?? "precise"),
                unit: Unit.fromName(try call.optionValue("unit", default: "auto") ?? "auto")
            )
            do {
                return try format(call.rawValue, options: options)
            } catch let error as MF2Error {
                guard error.code == "bad-operand", let sourceValue = relativeTimeSourceValue(call.inheritedSource) else {
                    throw error
                }
                return try format(.string(sourceValue), options: options)
            }
        }

        private func relativeTimeSourceValue(_ source: MF2FunctionSource?) -> String? {
            var current = source
            while let source = current {
                if isRelativeTimeNumericSourceFunction(source.function.name) {
                    return source.value
                }
                current = source.inheritedSource
            }
            return nil
        }

        private func isRelativeTimeNumericSourceFunction(_ name: String) -> Bool {
            switch name {
            case "number", "integer", "percent", "offset", "currency", "relativeTime":
                return true
            default:
                return false
            }
        }

        private func relativeTerm(
            locale: String,
            style: Style,
            unit: Unit,
            offset: String
        ) throws -> String? {
            try unitData(locale: locale, style: style, unit: unit).relative[offset]
        }

        private func relativeTimePattern(
            data: UnitData,
            locale: String,
            style: Style,
            unit: Unit,
            direction: String,
            category: String
        ) throws -> String {
            let patterns = direction == "past" ? data.past : data.future
            if let pattern = patterns[category] ?? patterns["other"] {
                return pattern
            }
            throw MF2Error.missingLocaleData(
                "Missing relative-time pattern for \(locale)/\(style.key)/\(unit.key)/\(direction)."
            )
        }

        private func unitData(locale: String, style: Style, unit: Unit) throws -> UnitData {
            let patternSet = try patternSetFor(locale: locale)
            guard let data = patternSet.data[style.key]?[unit.key] else {
                throw MF2Error.missingLocaleData("Missing relative-time unit data for \(locale)/\(style.key)/\(unit.key).")
            }
            return data
        }

        private func patternSetFor(locale: String) throws -> PatternSet {
            for candidate in MF2LocaleKey.lookupChain(locale) {
                guard let setId = localeMap[candidate] else {
                    continue
                }
                guard let patternSet = patternSets[setId] else {
                    throw MF2Error.missingLocaleData("Missing relative-time pattern set \(setId).")
                }
                return patternSet
            }
            throw MF2Error.missingLocaleData("Missing relative-time locale data for \(locale).")
        }
    }

    public static func registry(data: Data) throws -> MF2FunctionRegistry {
        try Formatter(data: data).registry()
    }

    public static func format(_ value: MF2Value, data: Data, options: Options = Options()) throws -> String {
        try Formatter(data: data).format(value, options: options)
    }

    private static func localeOption(_ locale: String) throws -> String {
        let value = locale.isEmpty ? "en" : locale
        if value.count > maxLocaleLength {
            throw MF2Error.badOption("locale must not exceed 256 characters.")
        }
        return value
    }

    private static func optionName(_ value: String, name: String) throws -> String {
        if value.count > maxOptionLength {
            throw MF2Error.badOption("\(name) must not exceed 256 characters.")
        }
        return value
    }

    public static func formatToParts(_ value: MF2Value, data: Data, options: Options = Options()) throws -> [MF2FormattedPart] {
        try Formatter(data: data).formatToParts(value, options: options)
    }

    struct PolicyStep {
        let upper: Double
        let unit: Unit
    }

    private static func parseFiniteNumber(_ value: MF2Value) throws -> Double {
        let raw: String
        switch value {
        case let .number(value), let .string(value):
            raw = value.trimmingCharacters(in: .whitespacesAndNewlines)
        case .bool, .null:
            throw MF2Error.badOperand("Relative-time core requires a finite numeric value.")
        }
        guard raw.count <= maxOperandLength,
            !raw.isEmpty,
            raw.range(of: decimalNumberPattern, options: .regularExpression) != nil,
            let parsed = Double(raw),
            parsed.isFinite
        else {
            throw MF2Error.badOperand("Relative-time core requires a finite numeric value.")
        }
        return parsed
    }

    private static func selectUnit(_ seconds: Double, policy: Policy) -> Unit {
        let absolute = abs(seconds)
        return policy.steps.first(where: { absolute < $0.upper })?.unit ?? .year
    }

    private static func relativeTimeQuantity(_ seconds: Double, unit: Unit) throws -> RelativeQuantity {
        let absolute = abs(seconds)
        if absolute == 0 {
            return RelativeQuantity(value: 0, text: "0")
        }
        let quantity = max(1, floor(absolute / unit.seconds + 0.5))
        if quantity > maxRelativeTimeQuantity {
            throw MF2Error.badOperand("Relative-time core quantity is outside the supported range.")
        }
        return RelativeQuantity(value: quantity, text: mf2FormatWholeDoubleNative(quantity))
    }

    private static func useRelativeZero(policy: Policy, numeric: Numeric, seconds: Double) -> Bool {
        policy == .chat && numeric == .auto && abs(seconds) < 45
    }

    private static func relativeOffset(_ seconds: Double, unit: Unit, quantity: RelativeQuantity) -> String? {
        if quantity.value == 0 {
            return "0"
        }
        if abs(seconds) != quantity.value * unit.seconds {
            return nil
        }
        return isNegativeRelativeTime(seconds) ? "-\(quantity.text)" : quantity.text
    }

    private static func isNegativeRelativeTime(_ seconds: Double) -> Bool {
        seconds < 0 || seconds.sign == .minus
    }

    private struct RelativeQuantity {
        let value: Double
        let text: String
    }
}

File: mf2/rust/mojito-mf2/src/relative_time_core.rs
use std::collections::BTreeMap;
use std::sync::Arc;

use serde::Deserialize;

use crate::cldr_plural_rules::{select_cardinal, NumberOperands};
use crate::diagnostic::Diagnostic;
use crate::formatter::{ArgumentValue, FormattedPart, FunctionCall, FunctionRegistry};
use crate::locale_key::locale_lookup_chain;

const DEFAULT_LOCALE: &str = "en";
const MAX_LOCALE_LENGTH: usize = 256;
const MAX_OPTION_LENGTH: usize = 256;
const MAX_OPERAND_LENGTH: usize = 256;
const MAX_RELATIVE_TIME_QUANTITY: f64 = 1_000_000_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RelativeTimeCoreStyle {
    Long,
    Short,
    Narrow,
}

impl RelativeTimeCoreStyle {
    pub fn from_name(value: &str) -> Result<Self, Diagnostic> {
        match option_name(value, "style")? {
            "long" => Ok(Self::Long),
            "short" => Ok(Self::Short),
            "narrow" => Ok(Self::Narrow),
            _ => Err(bad_option("style must be one of long, short, narrow.")),
        }
    }

    fn key(self) -> &'static str {
        match self {
            Self::Long => "long",
            Self::Short => "short",
            Self::Narrow => "narrow",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RelativeTimeCoreNumeric {
    Always,
    Auto,
}

impl RelativeTimeCoreNumeric {
    pub fn from_name(value: &str) -> Result<Self, Diagnostic> {
        match option_name(value, "numeric")? {
            "always" => Ok(Self::Always),
            "auto" => Ok(Self::Auto),
            _ => Err(bad_option("numeric must be one of always, auto.")),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RelativeTimeCorePolicy {
    Precise,
    Compact,
    Chat,
}

impl RelativeTimeCorePolicy {
    pub fn from_name(value: &str) -> Result<Self, Diagnostic> {
        match option_name(value, "policy")? {
            "precise" => Ok(Self::Precise),
            "compact" => Ok(Self::Compact),
            "chat" => Ok(Self::Chat),
            _ => Err(bad_option("policy must be one of precise, compact, chat.")),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RelativeTimeCoreUnit {
    Auto,
    Second,
    Minute,
    Hour,
    Day,
    Week,
    Month,
    Quarter,
    Year,
}

impl RelativeTimeCoreUnit {
    pub fn from_name(value: &str) -> Result<Self, Diagnostic> {
        match option_name(value, "unit")? {
            "auto" => Ok(Self::Auto),
            "second" => Ok(Self::Second),
            "minute" => Ok(Self::Minute),
            "hour" => Ok(Self::Hour),
            "day" => Ok(Self::Day),
            "week" => Ok(Self::Week),
            "month" => Ok(Self::Month),
            "quarter" => Ok(Self::Quarter),
            "year" => Ok(Self::Year),
            _ => Err(bad_option(
                "unit must be one of auto, second, minute, hour, day, week, month, quarter, year.",
            )),
        }
    }

    fn key(self) -> &'static str {
        match self {
            Self::Auto => "auto",
            Self::Second => "second",
            Self::Minute => "minute",
            Self::Hour => "hour",
            Self::Day => "day",
            Self::Week => "week",
            Self::Month => "month",
            Self::Quarter => "quarter",
            Self::Year => "year",
        }
    }

    fn seconds(self) -> f64 {
        match self {
            Self::Auto => 1.0,
            Self::Second => 1.0,
            Self::Minute => 60.0,
            Self::Hour => 3_600.0,
            Self::Day => 86_400.0,
            Self::Week => 604_800.0,
            Self::Month => 2_592_000.0,
            Self::Quarter => 7_776_000.0,
            Self::Year => 31_536_000.0,
        }
    }
}

#[derive(Debug, Clone)]
pub struct RelativeTimeCoreOptions {
    pub locale: String,
    pub style: RelativeTimeCoreStyle,
    pub numeric: RelativeTimeCoreNumeric,
    pub policy: RelativeTimeCorePolicy,
    pub unit: RelativeTimeCoreUnit,
}

impl Default for RelativeTimeCoreOptions {
    fn default() -> Self {
        Self {
            locale: DEFAULT_LOCALE.to_string(),
            style: RelativeTimeCoreStyle::Short,
            numeric: RelativeTimeCoreNumeric::Always,
            policy: RelativeTimeCorePolicy::Precise,
            unit: RelativeTimeCoreUnit::Auto,
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct RelativeTimeCoreData {
    #[serde(rename = "localeMap")]
    pub locale_map: BTreeMap<String, String>,
    #[serde(rename = "patternSets")]
    pub pattern_sets: Vec<RelativeTimeCorePatternSet>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RelativeTimeCorePatternSet {
    pub id: String,
    pub data: BTreeMap<String, BTreeMap<String, RelativeTimeCoreUnitData>>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RelativeTimeCoreUnitData {
    #[serde(default)]
    pub future: BTreeMap<String, String>,
    #[serde(default)]
    pub past: BTreeMap<String, String>,
    #[serde(default)]
    pub relative: BTreeMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct RelativeTimeCoreFormatter {
    locale_map: BTreeMap<String, String>,
    pattern_sets: BTreeMap<String, RelativeTimeCorePatternSet>,
}

impl RelativeTimeCoreFormatter {
    pub fn new(data: RelativeTimeCoreData) -> Result<Self, Diagnostic> {
        if data.locale_map.is_empty() || data.pattern_sets.is_empty() {
            return Err(missing_locale_data(
                "Relative-time core data has an unsupported shape.",
            ));
        }
        let pattern_sets: BTreeMap<_, _> = data
            .pattern_sets
            .into_iter()
            .filter(|pattern_set| !pattern_set.id.is_empty() && !pattern_set.data.is_empty())
            .map(|pattern_set| (pattern_set.id.clone(), pattern_set))
            .collect();
        if pattern_sets.is_empty() {
            return Err(missing_locale_data(
                "Relative-time core data has an unsupported shape.",
            ));
        }
        Ok(Self {
            locale_map: data.locale_map,
            pattern_sets,
        })
    }

    pub fn format(
        &self,
        value: impl Into<ArgumentValue>,
        options: &RelativeTimeCoreOptions,
    ) -> Result<String, Diagnostic> {
        let seconds = parse_finite_number(&value.into())?;
        let unit = if options.unit == RelativeTimeCoreUnit::Auto {
            select_unit(seconds, options.policy)
        } else {
            options.unit
        };
        let quantity = quantity(seconds, unit)?;
        let locale = locale_option(&options.locale)?;

        if use_relative_zero(options.policy, options.numeric, seconds) {
            if let Some(relative) = self.relative_term(locale, options.style, unit, "0")? {
                return Ok(relative.to_string());
            }
        }
        if options.numeric == RelativeTimeCoreNumeric::Auto {
            if let Some(offset) = relative_offset(seconds, unit, quantity) {
                if let Some(relative) = self.relative_term(locale, options.style, unit, &offset)? {
                    return Ok(relative.to_string());
                }
            }
        }

        let direction = if is_negative_relative_time(seconds) {
            "past"
        } else {
            "future"
        };
        let category = select_cardinal(
            locale,
            NumberOperands::from_str(&quantity.to_string())
                .expect("integer quantity is plural operand"),
        );
        let pattern =
            self.relative_time_pattern(locale, options.style, unit, direction, category)?;
        Ok(pattern.replace("{0}", &quantity.to_string()))
    }

    pub fn format_to_parts(
        &self,
        value: impl Into<ArgumentValue>,
        options: &RelativeTimeCoreOptions,
    ) -> Result<Vec<FormattedPart>, Diagnostic> {
        Ok(vec![FormattedPart::Text {
            value: self.format(value, options)?,
        }])
    }

    pub fn into_function_registry(self) -> FunctionRegistry {
        let formatter = Arc::new(self);
        FunctionRegistry::portable()
            .with_function("relativeTime", move |call| formatter.format_call(call))
    }

    fn format_call(&self, call: FunctionCall<'_>) -> Result<String, Diagnostic> {
        let mut options = RelativeTimeCoreOptions {
            locale: call.locale().to_string(),
            ..RelativeTimeCoreOptions::default()
        };
        if let Some(value) = call.option_value("style")? {
            options.style = RelativeTimeCoreStyle::from_name(&value)?;
        }
        if let Some(value) = call.option_value("numeric")? {
            options.numeric = RelativeTimeCoreNumeric::from_name(&value)?;
        }
        if let Some(value) = call.option_value("policy")? {
            options.policy = RelativeTimeCorePolicy::from_name(&value)?;
        }
        if let Some(value) = call.option_value("unit")? {
            options.unit = RelativeTimeCoreUnit::from_name(&value)?;
        }
        self.format(call.raw_value().clone(), &options)
    }

    fn relative_term(
        &self,
        locale: &str,
        style: RelativeTimeCoreStyle,
        unit: RelativeTimeCoreUnit,
        offset: &str,
    ) -> Result<Option<&str>, Diagnostic> {
        Ok(self
            .unit_data(locale, style, unit)?
            .relative
            .get(offset)
            .map(String::as_str))
    }

    fn relative_time_pattern(
        &self,
        locale: &str,
        style: RelativeTimeCoreStyle,
        unit: RelativeTimeCoreUnit,
        direction: &str,
        category: &str,
    ) -> Result<&str, Diagnostic> {
        let unit_data = self.unit_data(locale, style, unit)?;
        let patterns = if direction == "past" {
            &unit_data.past
        } else {
            &unit_data.future
        };
        patterns
            .get(category)
            .or_else(|| patterns.get("other"))
            .map(String::as_str)
            .ok_or_else(|| {
                missing_locale_data(format!(
                    "Missing relative-time pattern for {locale}/{}/{}/{direction}.",
                    style.key(),
                    unit.key()
                ))
            })
    }

    fn unit_data(
        &self,
        locale: &str,
        style: RelativeTimeCoreStyle,
        unit: RelativeTimeCoreUnit,
    ) -> Result<&RelativeTimeCoreUnitData, Diagnostic> {
        let pattern_set = self.pattern_set_for(locale)?;
        pattern_set
            .data
            .get(style.key())
            .and_then(|style_data| style_data.get(unit.key()))
            .ok_or_else(|| {
                missing_locale_data(format!(
                    "Missing relative-time unit data for {locale}/{}/{}.",
                    style.key(),
                    unit.key()
                ))
            })
    }

    fn pattern_set_for(&self, locale: &str) -> Result<&RelativeTimeCorePatternSet, Diagnostic> {
        for candidate in locale_lookup_chain(locale) {
            let Some(set_id) = self.locale_map.get(&candidate) else {
                continue;
            };
            return self.pattern_sets.get(set_id).ok_or_else(|| {
                missing_locale_data(format!("Missing relative-time pattern set {set_id}."))
            });
        }
        Err(missing_locale_data(format!(
            "Missing relative-time locale data for {locale}."
        )))
    }
}

pub fn relative_time_core_function_registry(
    data: RelativeTimeCoreData,
) -> Result<FunctionRegistry, Diagnostic> {
    Ok(RelativeTimeCoreFormatter::new(data)?.into_function_registry())
}

pub fn format_relative_time_core(
    value: impl Into<ArgumentValue>,
    data: &RelativeTimeCoreData,
    options: &RelativeTimeCoreOptions,
) -> Result<String, Diagnostic> {
    RelativeTimeCoreFormatter::new(data.clone())?.format(value, options)
}

pub fn format_relative_time_core_to_parts(
    value: impl Into<ArgumentValue>,
    data: &RelativeTimeCoreData,
    options: &RelativeTimeCoreOptions,
) -> Result<Vec<FormattedPart>, Diagnostic> {
    RelativeTimeCoreFormatter::new(data.clone())?.format_to_parts(value, options)
}

fn locale_option(locale: &str) -> Result<&str, Diagnostic> {
    let value = if locale.is_empty() {
        DEFAULT_LOCALE
    } else {
        locale
    };
    if value.chars().count() > MAX_LOCALE_LENGTH {
        return Err(bad_option("locale must not exceed 256 characters."));
    }
    Ok(value)
}

fn option_name<'a>(value: &'a str, name: &str) -> Result<&'a str, Diagnostic> {
    if value.chars().count() > MAX_OPTION_LENGTH {
        return Err(bad_option(format!(
            "{name} must not exceed 256 characters."
        )));
    }
    Ok(value)
}

fn parse_finite_number(value: &ArgumentValue) -> Result<f64, Diagnostic> {
    let text = match value {
        ArgumentValue::Number(value) | ArgumentValue::String(value) => value.trim(),
        ArgumentValue::Bool(_) | ArgumentValue::Null => {
            return Err(bad_operand(
                "Relative-time core requires a finite numeric value.",
            ));
        }
    };
    if text.is_empty() {
        return Err(bad_operand(
            "Relative-time core requires a finite numeric value.",
        ));
    }
    if text.chars().count() > MAX_OPERAND_LENGTH {
        return Err(bad_operand(
            "Relative-time core requires a finite numeric value.",
        ));
    }
    if !is_decimal_number(text) {
        return Err(bad_operand(
            "Relative-time core requires a finite numeric value.",
        ));
    }
    let parsed = text
        .parse::<f64>()
        .map_err(|_| bad_operand("Relative-time core requires a finite numeric value."))?;
    if parsed.is_finite() {
        Ok(parsed)
    } else {
        Err(bad_operand(
            "Relative-time core requires a finite numeric value.",
        ))
    }
}

fn is_decimal_number(text: &str) -> bool {
    let bytes = text.as_bytes();
    let mut index = 0;
    if matches!(bytes.first(), Some(b'-')) {
        index += 1;
    }

    match bytes.get(index) {
        Some(b'0') => {
            index += 1;
            if matches!(bytes.get(index), Some(b'0'..=b'9')) {
                return false;
            }
        }
        Some(b'1'..=b'9') => {
            index += 1;
            while matches!(bytes.get(index), Some(b'0'..=b'9')) {
                index += 1;
            }
        }
        _ => return false,
    }

    if index < bytes.len() && bytes[index] == b'.' {
        index += 1;
        let fraction_start = index;
        while index < bytes.len() && bytes[index].is_ascii_digit() {
            index += 1;
        }
        if index == fraction_start {
            return false;
        }
    }

    if index < bytes.len() && matches!(bytes[index], b'e' | b'E') {
        index += 1;
        if index < bytes.len() && matches!(bytes[index], b'+' | b'-') {
            index += 1;
        }
        let exponent_start = index;
        while index < bytes.len() && bytes[index].is_ascii_digit() {
            index += 1;
        }
        if index == exponent_start {
            return false;
        }
    }

    index == bytes.len()
}

fn select_unit(seconds: f64, policy: RelativeTimeCorePolicy) -> RelativeTimeCoreUnit {
    let absolute = seconds.abs();
    let steps: &[(f64, RelativeTimeCoreUnit)] = match policy {
        RelativeTimeCorePolicy::Precise => &[
            (60.0, RelativeTimeCoreUnit::Second),
            (3_600.0, RelativeTimeCoreUnit::Minute),
            (86_400.0, RelativeTimeCoreUnit::Hour),
            (604_800.0, RelativeTimeCoreUnit::Day),
            (2_592_000.0, RelativeTimeCoreUnit::Week),
            (31_536_000.0, RelativeTimeCoreUnit::Month),
            (f64::INFINITY, RelativeTimeCoreUnit::Year),
        ],
        RelativeTimeCorePolicy::Compact => &[
            (60.0, RelativeTimeCoreUnit::Second),
            (3_600.0, RelativeTimeCoreUnit::Minute),
            (86_400.0, RelativeTimeCoreUnit::Hour),
            (f64::INFINITY, RelativeTimeCoreUnit::Day),
        ],
        RelativeTimeCorePolicy::Chat => &[
            (45.0, RelativeTimeCoreUnit::Second),
            (2_700.0, RelativeTimeCoreUnit::Minute),
            (79_200.0, RelativeTimeCoreUnit::Hour),
            (604_800.0, RelativeTimeCoreUnit::Day),
            (f64::INFINITY, RelativeTimeCoreUnit::Week),
        ],
    };
    steps
        .iter()
        .find(|(upper, _)| absolute < *upper)
        .map(|(_, unit)| *unit)
        .unwrap_or(RelativeTimeCoreUnit::Year)
}

fn quantity(seconds: f64, unit: RelativeTimeCoreUnit) -> Result<i64, Diagnostic> {
    let absolute = seconds.abs();
    if absolute == 0.0 {
        return Ok(0);
    }
    let quantity = ((absolute / unit.seconds()) + 0.5).floor().max(1.0);
    if quantity > MAX_RELATIVE_TIME_QUANTITY {
        return Err(bad_operand(
            "Relative-time core quantity is outside the supported range.",
        ));
    }
    Ok(quantity as i64)
}

fn use_relative_zero(
    policy: RelativeTimeCorePolicy,
    numeric: RelativeTimeCoreNumeric,
    seconds: f64,
) -> bool {
    policy == RelativeTimeCorePolicy::Chat
        && numeric == RelativeTimeCoreNumeric::Auto
        && seconds.abs() < 45.0
}

fn relative_offset(seconds: f64, unit: RelativeTimeCoreUnit, quantity: i64) -> Option<String> {
    if quantity == 0 {
        return Some("0".to_string());
    }
    if seconds.abs() != quantity as f64 * unit.seconds() {
        return None;
    }
    Some(if is_negative_relative_time(seconds) {
        format!("-{quantity}")
    } else {
        quantity.to_string()
    })
}

fn is_negative_relative_time(seconds: f64) -> bool {
    seconds.is_sign_negative()
}

fn bad_operand(message: impl Into<String>) -> Diagnostic {
    Diagnostic::new("bad-operand", message.into(), 0, 0)
}

fn bad_option(message: impl Into<String>) -> Diagnostic {
    Diagnostic::new("bad-option", message.into(), 0, 0)
}

fn missing_locale_data(message: impl Into<String>) -> Diagnostic {
    Diagnostic::new("missing-locale-data", message.into(), 0, 0)
}
